lib/spack/spack/compilers/__init__.py

```diff
@@ -255,41 +255,69 @@ def compilers_for_arch(arch_spec, scope=None):
     return list(get_compilers(config, arch_spec=arch_spec))
 
 
-def compiler_from_config_entry(items):
-    config_id = id(items)
-    compiler = _compiler_cache.get(config_id, None)
+class StrongReference(object):
+    """This acts as a hashable reference to any object (regardless of whether
+       the object itself is hashable) and also prevents the object from being
+       garbage-collected (so if two StrongReference objects are equal, they
```

Comment:
IMHO I think the description here could be simplified.  For example, something along the lines:

`This class provides hashable references to any object and, e.g., when paired with a global dictionary, prevents the associated object from being garbage-collected.`